### PR TITLE
Removing Macros + Some endpoints support multiple instantiations of themselves

### PR DESCRIPTION
**Pull request by sagark**
- Replace macros with structs (https://github.com/ucb-bar/midas/pull/96)
    - Adds a macro used in firesim_top.cc to create a struct
    - Struct then gets passed into the constructor, provides access to mmio reg addrs
    - endpoint destructor responsible for free()ing struct
- Added TODO list for some missing things that will be in a separate PR
- Start replacing arg parsing for certain endpoints to support multiple instantiations of that endpoint (tracer, NIC, blkdev) in prep for supernode
- Some endpoints need more work to support multiple instantiations (Serial (arg parsing), UART (pty support))

### CODE REVIEW

File: sim/src/main/cc/endpoints/blockdev.cc [hunk 1]
```diff
@@ -92,13 +94,13 @@ blockdev_t::~blockdev_t() {
  * Here, we set control regs e.g. for # sectors, allowed request length
  * at boot */
 void blockdev_t::init() {
-#ifdef BLOCKDEVWIDGET_0
+#ifdef BLOCKDEVWIDGET_struct_guard
```
Comment (davidbiancolin) [inline sim/src/main/cc/endpoints/blockdev.cc]: Can't we just remove all this and just not instantiate the endpoint now?

Comment (davidbiancolin) [reply sim/src/main/cc/endpoints/blockdev.cc]: By all this, i mean ifdefs around function bodies.

**Comment by sagark** [comment]
@davidbiancolin this is ready for review now. This doesn't solve all problems (namely integrating other constants the widget chisel writer might want to emit), but I'd like to get it merged in since it's a lot of stuff that easily causes merge conflicts.

File: sim/src/main/cc/endpoints/simplenic.cc [hunk 2]
```diff
@@ -79,34 +89,34 @@ simplenic_t::simplenic_t(
...
                 }
             } else {
-                fprintf(stderr, "INVALID MAC ADDRESS SUPPLIED WITH +macaddr=\n");
+                fprintf(stderr, "INVALID MAC ADDRESS SUPPLIED WITH +macaddrN=\n");
             }
         }
-        if (arg.find("+netbw=") == 0) {
-            char *str = const_cast<char*>(arg.c_str()) + 7;
+        if (arg.find(netbw_arg) == 0) {
+            char *str = const_cast<char*>(arg.c_str()) + netbw_arg.length();
```
Comment (davidbiancolin) [inline sim/src/main/cc/endpoints/simplenic.cc]: Tight. That old pattern had to die.

**Comment by sagark** [comment]
yeah, that's right

**Pull request merged**